From 78133be82350eabd3e4668d9877bc1898da1bffe Mon Sep 17 00:00:00 2001
From: CL Wang <cl634@andestech.com>
Date: Wed, 30 Oct 2024 17:40:00 +0800
Subject: [PATCH 131/168] ALSA: andes: ftssp010: Assign FTSSP to use DMA
 channel 7 (#218)

Reserve DMA channel 1 for the dmatest mentioned in the LDUM document.
After system startup, the DMA channel usage is as follows:

Channel 0 => MMC
Channel 1 => LDUM dmatest
...
Channel 7 => FTSSP

Reviewed-on: https://gitea.andestech.com/RD-SW/linux/pulls/218
Reviewed-by: randolph <randolph@andestech.com>
Reviewed-by: Locus Wei-Han Chen <locus84@andestech.com>
Reviewed-by: Leo Yu-Chi Liang <ycliang@andestech.com>
Reviewed-by: Tim Shih-Ting OuYang <tim609@andestech.com>
Co-authored-by: CL Wang <cl634@andestech.com>
Co-committed-by: CL Wang <cl634@andestech.com>
Co-authored-by: CL Wang <cl634@andestech.com>
Co-committed-by: CL Wang <cl634@andestech.com>
---
 sound/v5/ftssp010_dma.c | 14 +++++++++++---
 1 file changed, 11 insertions(+), 3 deletions(-)

diff --git a/sound/v5/ftssp010_dma.c b/sound/v5/ftssp010_dma.c
index 8d7b0fc258d4..0a89f737b2fc 100644
--- a/sound/v5/ftssp010_dma.c
+++ b/sound/v5/ftssp010_dma.c
@@ -25,8 +25,9 @@
 #include <linux/of_platform.h>
 #include "ftssp010_dma.h"
 
-#define DMAD_DRB_POOL_SIZE    32
-#define DMAD_AHB_MAX_CHANNELS DMAC_MAX_CHANNELS
+#define FTSSP_DMA_CH		"dma0chan7"
+#define DMAD_DRB_POOL_SIZE	32
+#define DMAD_AHB_MAX_CHANNELS	DMAC_MAX_CHANNELS
 
 enum DMAD_DRQ_FLAGS {
 	DMAD_DRQ_STATE_READY = 0x00000001,	/* channel allocation status */
@@ -2235,6 +2236,13 @@ int dmad_drain_requests(dmad_chreq *ch_req, u8 shutdown)
 	return dmad_channel_drain(ch_req->controller, ch_req->drq, shutdown);
 }
 
+static bool dmad_chan_filter(struct dma_chan *chan, void *param)
+{
+	char *chan_name = (char *)param;
+
+	return strcmp(dma_chan_name(chan), chan_name) == 0;
+}
+
 int dmad_init(void)
 {
 	struct device_node *dmac = NULL;
@@ -2245,7 +2253,7 @@ int dmad_init(void)
 	dma_cap_zero(mask);
 	dma_cap_set(DMA_CYCLIC, mask);
 	dma_cap_set(DMA_SLAVE, mask);
-	chan = dma_request_channel(mask, NULL, NULL);
+	chan = dma_request_channel(mask, dmad_chan_filter, FTSSP_DMA_CH);
 	if (chan == NULL) {
 		pr_err("Failed to allocate a DMA channel for SSP.\n");
 		return -EINVAL;
-- 
2.34.1

